url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/coffee/ShowCoffee.java

```diff
@@ -0,0 +1,46 @@
+package sjm.examples.book.coffee;
+
+import java.io.*;
+
+import sjm.parse.*;
+import sjm.parse.tokens.*;
+
+/*
+ * Copyright (c) 2000 Steven J. Metsker. All Rights Reserved.
+ * 
+ * Steve Metsker makes no representations or warranties about
+ * the fitness of this software for any particular purpose, 
+ * including the implied warranty of merchantability.
+ */
+
+/**
+ * Show the recognition of a list of types of coffee, 
+ * reading from a file.
+ *
+ * @author Steven J. Metsker
+ *
+ * @version 1.0 
+ */
+public class ShowCoffee {
+	/**
+	 * Show how to recognize coffees in a file.
+	 */
+	public static void main(String args[]) throws Exception {
+
+		InputStream is = ClassLoader.getSystemResourceAsStream("coffee.txt");
+		BufferedReader r = new BufferedReader(new InputStreamReader(is));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *r*. Execution paths that do not contain closure statements were detected. Ensure *r* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).